### PR TITLE
Telemetry

### PR DESCRIPTION
**Pull request by SCdF**
Provides a simple interface with one implementation of telemetry recording and aggregation.

medic/medic-webapp#2031

# Review checklist

- [ ] Readable: Concise, well named, follows the [style guide](https://github.com/medic/medic-docs/blob/master/development/style-guide.md), documented if necessary.
- [ ] Documented: Announced in Changes.md in plain English. Configuration and user documentation on [medic-docs](https://github.com/medic/medic-docs/)
- [ ] Tested: Unit and/or e2e where appropriate
- [ ] Internationalised: All user facing text
- [ ] Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in Changes.md.

### CODE REVIEW

File: webapp/src/js/services/db.js [hunk 1]
```diff
@@ -87,5 +87,4 @@ angular.module('inboxServices').factory('DB',
     }
 
     return get;
-  }
-);
```
Comment (garethbowen) [inline webapp/src/js/services/db.js]: Consider reverting this file - nothing actually changed, right?

File: webapp/src/js/services/telemetry.js [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+        screen: {
+          width: $window.screen.availWidth,
+          height: $window.screen.availHeight,
+        },
+        // REVIEWER: while we can expand this in the future are their other
+        // useful things you would like to see here?
+        //
+        // REVIEWER: do we want to ship with extra data pulled from Android via
+        // medic-android? This would allow us to know true android version, disk
+        // space, stuff like that.
```
Comment (garethbowen) [inline webapp/src/js/services/telemetry.js]: Yeah I think the android information is worth pulling, particularly with the performance data.

File: webapp/src/js/services/telemetry.js [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+    var generateMetadataSection = function() {
+      var date = moment(getLastAggregatedDate());
+      return {
+        year: date.year(),
+        month: date.month(),
+        user: Session.userCtx().name,
+        deviceId: getUniqueDeviceId(),
+        // REVIEWER: is there anything else we think we should store here?
+        // Candidates would be generic metadata that is not directly device
+        // related or DB related
```
Comment (garethbowen) [inline webapp/src/js/services/telemetry.js]: None I can think of.

File: webapp/src/js/services/telemetry.js [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+      };
+    };
+
+    var generateDeviceStats = function() {
+      return {
+        userAgent: $window.navigator.userAgent,
+        language: $window.navigator.language,
```
Comment (garethbowen) [inline webapp/src/js/services/telemetry.js]: Language is probably not very useful - I think we ignore it anyway, and resort to the language available in the user-settings doc that we can query from couchdb directly.

File: webapp/src/js/services/telemetry.js [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+        hardwareConcurrency: $window.navigator.hardwareConcurrency,
+        screen: {
+          width: $window.screen.availWidth,
+          height: $window.screen.availHeight,
+        },
+        // REVIEWER: while we can expand this in the future are their other
+        // useful things you would like to see here?
```
Comment (garethbowen) [inline webapp/src/js/services/telemetry.js]: LGTM.

File: webapp/src/js/services/telemetry.js [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+        ])
+        .then(function(qAll) {
+          var reduceResult = qAll[0];
+          var infoResult = qAll[1];
+
+          var aggregateDoc = {
+            type: 'telemetry',
+          };
+
+          // REVIEWER: is there a better name than 'stats' for aggregated telemetry records?
```
Comment (garethbowen) [inline webapp/src/js/services/telemetry.js]: What about `metrics`? IDK...

File: webapp/src/js/app.js [hunk 7]
```diff
@@ -1,3 +1,10 @@
+// While we already do this earlier in inbox.js we have to check again for Karma
+// tests as they don't hit that code
+if (!window.startupTimes) {
+  window.startupTimes = {};
+}
+window.startupTimes.firstCodeExecution = performance.now();
```
Comment (kennsippell) [inline webapp/src/js/app.js]: You might consider putting this in a <script> tag inside the `<head>` of `inbox.html` so that it includes time to download and parse all assets.

File: webapp/src/js/services/db.js (hunk 1 above)
Comment (SCdF) [reply webapp/src/js/services/db.js]: Yeah! I wanted to at the time but my energy drained when I realised to do it I either had to:
 - Rebase from the start and try to edit out all the changes
 - Revert it, disable prettier and *then* commit the revert, because otherwise prettier will pretty up my de-prettying.

I'll work it out and fix it.

File: webapp/src/js/services/telemetry.js [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+  .module('inboxServices')
+  .factory('Telemetry', function($log, $q, $window, DB, Session) {
+    'use strict';
+    'ngInject';
+
+    var queue = $q.resolve();
+
+    // Intentionally scoped to the whole browser (for this domain). We can then
+    // tell if multiple users use the same device
+    var DEVICE_ID_KEY = 'medic-telemetry-device-id';
```
Comment (alxndrsn) [inline webapp/src/js/services/telemetry.js]: I'm not sure if the current implementation of
```
        uniqueDeviceId = uuid();
        $window.localStorage.setItem(DEVICE_ID_KEY, uniqueDeviceId);
```
will actually be shared by multiple users using the same device in practice.  Mobile users will "log out" to change user by clearing the device cache, which will presumably clear this `localStorage` value.  On desktop, if I use Chrome and then Firefox, the device ID would be different for each of them - despite being the same device.  

There are definitely more reliable ways we could fingerprint an Android device, e.g. by requesting info from Java.  Alternatively, renaming the "device id" to something more accurately reflecting its limitations could help communicate what this actually represents.

File: webapp/src/js/services/telemetry.js [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+      'medic',
+      Session.userCtx().name,
+      'telemetry-date',
+    ].join('-');
+
+    var getDb = function() {
+      var dbName = $window.localStorage.getItem(DB_ID_KEY);
+      if (!dbName) {
+        dbName =
+          'medic-user-' + Session.userCtx().name + '-telemetry-' + uuid();
```
Comment (alxndrsn) [inline webapp/src/js/services/telemetry.js]: What's the benefit of having a UUID here?  Could you just use `medic-${userName}-telemetry` as the db name directly and achieve the same?

File: webapp/src/js/services/telemetry.js [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+
+    // Intentionally scoped to the specific user, as they may perform a
+    // different role (online vs. offline being being the most obvious) with
+    // different performance implications
+    var DB_ID_KEY = ['medic', Session.userCtx().name, 'telemetry-db'].join('-');
+    var LAST_AGGREGATED_DATE_KEY = [
+      'medic',
+      Session.userCtx().name,
+      'telemetry-date',
+    ].join('-');
```
Comment (alxndrsn) [inline webapp/src/js/services/telemetry.js]: Rather than generating multiple user-specific keys for local storage and using these to persist primitive types, might it be simpler to have a single user-specific key and persist an object?  E.g.
```
localStorage.getItem(`medic-${userName}-telemetry-metadata`)
```
returning
```
{
  db_id: 'medic-user-myUsername-abc-123`,
  last_aggregated_date: 1234567890,
  some_key_added_in_the_future: 'value',
}
```
This question may be moot if a UUID is not required for the user's db (see below).

File: webapp/src/js/services/telemetry.js [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+    };
+
+    var getLastAggregatedDate = function() {
+      var date = parseInt(
+        $window.localStorage.getItem(LAST_AGGREGATED_DATE_KEY)
+      );
+      if (!date) {
+        date = Date.now();
+        $window.localStorage.setItem(LAST_AGGREGATED_DATE_KEY, date);
+      }
```
Comment (alxndrsn) [inline webapp/src/js/services/telemetry.js]: It makes sense to me that `DEVICE_ID_KEY` might be lazily initialised in the getter, but less so for the last aggregated date.  Also I can't see any other place where `LAST_AGGREGATED_DATE_KEY` gets set, so this would have the value of the first call to `getLastAggregatedDate()` (although I haven't quite reasoned what this would be).

I think this should probably be set in the `aggregate()` function, and _only there_.

File: webapp/src/js/services/telemetry.js [hunk 12]
```diff
@@ -0,0 +1,215 @@
...
+    var storeConflictedAggregate = function(aggregateDoc) {
+      aggregateDoc.metadata.conflicted = true;
+      aggregateDoc._id = [aggregateDoc._id, 'conflicted', Date.now()].join('-');
+      return DB().put(aggregateDoc);
+    };
+
+    var aggregate = function(db) {
+      return $q
+        .all([
+          db.query(
```
Comment (alxndrsn) [inline webapp/src/js/services/telemetry.js]: Are there benefits or disadvantages of using a function here instead of a named view?

Comment (alxndrsn) [inline webapp/src/js/services/telemetry.js]: Could this be implemented as a Mango query?  I'm not familiar with them, but the Pouch docs suggest they should be preferred where possible.  OTOH I suspect Mango would be a bit slower, if it just ends up generating the `map` and `reduce` functions for us before executing them the same way.

File: webapp/src/js/services/telemetry.js [hunk 13]
```diff
@@ -0,0 +1,215 @@
...
+      // Specifically, a unique key that will be aggregated against, and if you
+      // are recording a timing (as opposed to an event) a value.
+      //
+      // While this function returns a promise, this is primarily for testing.
+      // It is not recommended you hold on to this promise and wait for it to
+      // resolve, for performance reasons.
+      //
+      // @param      {String}   key     a unique key that will be aggregated
+      //                                against later
+      // @param      {Number}   value   number to be aggregated. Defaults to 1.
```
Comment (alxndrsn) [inline webapp/src/js/services/telemetry.js]: I can't see code that would make `value` default to `1`.

File: webapp/src/js/services/telemetry.js [hunk 14]
```diff
@@ -0,0 +1,215 @@
...
+            type: 'telemetry',
+          };
+
+          aggregateDoc.metrics = convertReduceToKeyValues(reduceResult);
+          aggregateDoc.metadata = generateMetadataSection();
+          aggregateDoc._id = generateAggregateDocId(aggregateDoc.metadata);
+          aggregateDoc.device = generateDeviceStats();
+          aggregateDoc.dbInfo = infoResult;
+
+          return DB()
```
Comment (alxndrsn) [inline webapp/src/js/services/telemetry.js]: For performance reasons would it be better to put the aggregate docs in the individual user's meta DB?  Presumably these docs are write-once and never relevant to other users?  I appreciate that this would add work on the server-side to collect all telemetry docs for different users.

File: webapp/src/js/services/telemetry.js [hunk 15]
```diff
@@ -0,0 +1,215 @@
+var moment = require('moment'),
+  uuid = require('uuid/v4');
+
+angular
+  .module('inboxServices')
+  .factory('Telemetry', function($log, $q, $window, DB, Session) {
+    'use strict';
+    'ngInject';
+
+    var queue = $q.resolve();
```
Comment (alxndrsn) [inline webapp/src/js/services/telemetry.js]: I don't understand if there are performance implications to keeping a queue of all telemetry promises for the duration of the browser session.  Hopefully there are not...

File: webapp/src/js/services/telemetry.js [hunk 16]
```diff
@@ -0,0 +1,215 @@
...
+        userAgent: $window.navigator.userAgent,
+        hardwareConcurrency: $window.navigator.hardwareConcurrency,
+        screen: {
+          width: $window.screen.availWidth,
+          height: $window.screen.availHeight,
+        },
+        // TODO: expose some device information in the medic-android wrapper
+        // and pull it in here. Device memory, phone type, stuff like that
+        //
+        // https://github.com/medic/medic-webapp/issues/4882
```
Comment (alxndrsn) [inline webapp/src/js/services/telemetry.js]: Are TODOs acceptable iff accompanied by ticket number?

File: webapp/src/js/services/telemetry.js [hunk 17]
```diff
@@ -0,0 +1,215 @@
...
+
+          var aggregateDoc = {
+            type: 'telemetry',
+          };
+
+          aggregateDoc.metrics = convertReduceToKeyValues(reduceResult);
+          aggregateDoc.metadata = generateMetadataSection();
+          aggregateDoc._id = generateAggregateDocId(aggregateDoc.metadata);
+          aggregateDoc.device = generateDeviceStats();
+          aggregateDoc.dbInfo = infoResult;
```
Comment (alxndrsn) [inline webapp/src/js/services/telemetry.js]: What useful information does the db info contain?

File: webapp/src/js/services/telemetry.js [hunk 18]
```diff
@@ -0,0 +1,215 @@
...
+
+    // This should never happen (famous last words..), because we should only
+    // generate a new document for every month, which is part of the _id.
+    var storeConflictedAggregate = function(aggregateDoc) {
+      aggregateDoc.metadata.conflicted = true;
+      aggregateDoc._id = [aggregateDoc._id, 'conflicted', Date.now()].join('-');
+      return DB().put(aggregateDoc);
+    };
+
+    var aggregate = function(db) {
```
Comment (alxndrsn) [inline webapp/src/js/services/telemetry.js]: I think I've worked out what this function does.  It might be helpful to link to the documentation for couchdb's built-in `_stats` reduce function, and give an example of aggregated data.  OTOH perhaps I should already understand it!  From what I understand `aggregate().metrics` should have value something like:
```
{
  metric_a:  { sum: 492, min: 123, max: 123, count: 4, sumsqr: 60516 },
  metric_b:  { sum: -16, min: -4, max: -4, count: 4, sumsqr: 64 }
}
```

File: webapp/tests/karma/unit/services/telemetry.js [hunk 19]
```diff
@@ -0,0 +1,215 @@
...
+
+  afterEach(function() {
+    clock.restore();
+  });
+
+  describe('Record', () => {
+    it('records a piece of telemetry', () => {
+      storageGetItem.withArgs('medic-greg-telemetry-db').returns('dbname');
+      storageGetItem
+        .withArgs('medic-greg-telemetry-date')
```
Comment (alxndrsn) [inline webapp/tests/karma/unit/services/telemetry.js]: I misread `greg-telemetry-date` as being the telemetry date in the Gregorian calendar :smile:

File: webapp/tests/karma/unit/services/telemetry.js [hunk 20]
```diff
@@ -0,0 +1,215 @@
...
+
+  beforeEach(() => {
+    module('inboxApp');
+    clock = sinon.useFakeTimers(NOW);
+
+    storageGetItem = sinon.stub();
+    storageSetItem = sinon.stub();
+    storageRemoveItem = sinon.stub();
+
+    pouchDb = {};
```
Comment (alxndrsn) [inline webapp/tests/karma/unit/services/telemetry.js]: I'd be tempted to use an in-memory pouchdb here.

File: webapp/tests/karma/unit/services/telemetry.js [hunk 21]
```diff
@@ -0,0 +1,215 @@
...
+
+      pouchDb.post = sinon.stub().resolves();
+
+      return service.record('test', 100).then(() => {
+        chai.expect(pouchDb.post.callCount).to.equal(1);
+        chai
+          .expect(pouchDb.post.args[0][0])
+          .to.deep.include({ key: 'test', value: 100 });
+        chai.expect(pouchDb.post.args[0][0].date_recorded).to.be.above(0);
+        chai.expect(storageGetItem.callCount).to.equal(2);
```
Comment (alxndrsn) [inline webapp/tests/karma/unit/services/telemetry.js]: Do we care how many times `localStorage.getItem()` is called, or is this testing an implementation detail?

File: webapp/tests/karma/unit/services/telemetry.js [hunk 22]
```diff
@@ -0,0 +1,215 @@
...
+    inject(_Telemetry_ => {
+      service = _Telemetry_;
+    });
+  });
+
+  afterEach(function() {
+    clock.restore();
+  });
+
+  describe('Record', () => {
```
Comment (alxndrsn) [inline webapp/tests/karma/unit/services/telemetry.js]: is this describing `Telemetry.record()`?

File: webapp/tests/karma/unit/services/telemetry.js [hunk 23]
```diff
@@ -0,0 +1,215 @@
...
+        chai.expect(aggregatedDoc.metrics).to.deep.equal({
+          foo: 'stats',
+          bar: 'more stats',
+        });
+        chai.expect(aggregatedDoc.type).to.equal('telemetry');
+        chai.expect(aggregatedDoc.metadata.year).to.equal(2018);
+        chai.expect(aggregatedDoc.metadata.month).to.equal(9);
+        chai.expect(aggregatedDoc.metadata.user).to.equal('greg');
+        chai.expect(aggregatedDoc.dbInfo).to.deep.equal({ some: 'stats' });
+        chai.expect(aggregatedDoc.device).to.deep.equal({
```
Comment (alxndrsn) [inline webapp/tests/karma/unit/services/telemetry.js]: Would it be clearer if the whole expected document could be defined in a single test?  That way the structure of the aggregated doc would be clear from a glance at the tests.  E.g.
```
chai.expected(aggregatedDoc).to.deeplyEqual({
  type: 'telemetry',
  metadata: {
    year: 2018,
    month: 9,
    user: 'greg',
  },
  dbInfo: { some:'stats' },
  device: {
    userAgent: 'Agent Smith',
    hardwareConcurrency: 4,
    screen: { width:768, height:1024 },
  },
});
```
I guess there may be parts of the doc you're not interested in, but IIRC we have a lib in use for asserting that.

File: webapp/tests/karma/unit/services/telemetry.js [hunk 24]
```diff
@@ -0,0 +1,215 @@
...
+          hardwareConcurrency: 4,
+          screen: {
+            width: 768,
+            height: 1024,
+          },
+        });
+      });
+    });
+
+    it('deals with 409s by making the ID unique and noting the conflict in the new document', () => {
```
Comment (alxndrsn) [inline webapp/tests/karma/unit/services/telemetry.js]: I guess we should all know by now that 409 is a conflict :slightly_frowning_face:

File: webapp/tests/karma/unit/controllers/inbox.js [hunk 25]
```diff
@@ -1,30 +1,28 @@
...
+    scope,
+    snackbar,
+    spyState,
+    stubModal,
+    dummyId = 'dummydummy',
+    RecurringProcessManager,
+    changes,
+    changesListener = {},
+    changesSpy,
+    session;
```
[the remark this answers is not in the document]
Comment (SCdF) [reply webapp/tests/karma/unit/controllers/inbox.js]: I agree. This is "prettier" at work.

File: webapp/tests/karma/unit/controllers/inbox.js [hunk 26]
```diff
@@ -78,30 +76,33 @@ describe('InboxCtrl controller', () => {
         spyState = {
           go: sinon.spy(),
           current: { name: 'my.state.is.great' },
-          includes: () => { return true; }
+          includes: () => {
```
[the remark this answers is not in the document]
Comment (SCdF) [reply webapp/tests/karma/unit/controllers/inbox.js]: I didn't change this, "prettier" did.

File: webapp/src/js/controllers/inbox.js [hunk 27]
```diff
@@ -52,6 +53,20 @@ var feedback = require('../modules/feedback'),
   ) {
     'ngInject';
 
+    window.startupTimes.angularBootstrapped = performance.now();
+    Telemetry.record(
+      'boot_time_1_to_first_code_execution',
```
[the remark this answers is not in the document]
Comment (SCdF) [reply webapp/src/js/controllers/inbox.js]: So the names here are basically `namespace_num_description`. The number is just so you know ordering of each section. There is a namespace, but the namespace doesn't use a period because I thought it would be weird to have a period in a JSON property name.

The current aggregation is super duper simple, it just works out `_stats` for each key, and so keys can't, at least as the core identifier, be namespaced specifically at that point. Any namespacing would just be convention.

It wouldn't be hard though, to provide a little lift to the convention and change the aggregation so that `boot_time.1_to_first_code_execution`, once aggregated, ended up being:

```json
{
	"boot_time": {
		"1_to_first_code_execution": "blah",
		"2_to_bootstrap": "blah"
	}
}
```

Is that what you're interested in here?

File: webapp/src/js/services/telemetry.js (hunk 15 above)
Comment (SCdF) [reply webapp/src/js/services/telemetry.js]: Yeah I wondered about this too. I think I resolved it in my brain as being OK, but I don't remember what my reasoning was. @garethbowen do you remember why we thought this was OK? And wouldn't, for example, leak memory because it has to endlessly build up a stack trace?

If we're unsure at all IMO we should change tack to being a different form of queuing, though I'm not sure what that would be, since we need to wait until each promise is done, so I think it has to be an endless chain? @alxndrsn what would another implementation here be?

To be super clear, we keep on reassigning `queue` on each call, so we are "dropping the head" from a memory allocation point of view.

File: webapp/src/js/services/telemetry.js (hunk 8 above)
Comment (SCdF) [reply webapp/src/js/services/telemetry.js]: > Mobile users will "log out" to change user by clearing the device cache, which will presumably clear this localStorage value. On desktop, if I use Chrome and then Firefox, the device ID would be different for each of them - despite being the same device.

Those two scenarios (clearing your cache completely, and using different browsers) will not be caught here no. However, if you just log out of your session and log back in as someone else using the same browser, e.g. at a facility or something, that would be caught.

We could use something from android, but then it wouldn't work outside of that, and ideally it would be consistent in shape (ie both should be UUIDs).

The core "real" problem we're solving here, is that if you are logged in on two different devices / browsers, then I don't want your aggregate telemetry documents to conflict with each other. It is icing on the cake if we could tie different users back to the same device / browser.

Do you think `sessionId` sounds more correct?

File: webapp/src/js/services/telemetry.js (hunk 10 above)
Comment (SCdF) [reply webapp/src/js/services/telemetry.js]: I think local storage only stores strings ("DomStrings" even, however that is different), so it sort of nudges you into this pattern. I agree though, that packaging them up into one object would be nicer.

File: webapp/src/js/services/telemetry.js (hunk 9 above)
Comment (SCdF) [reply webapp/src/js/services/telemetry.js]: You know that's a great question. This may have been left over from testing when we were not deleting the old DBs. I don't remember if we still need it though, @garethbowen can you remember? It may be that `db.destroy()` has not completely cleaned up by the time you go to instantiate another DB of the same name in its place and it breaks horribly, but I think no longer using `angular-pouchdb` was all we really needed to do?

NB: about ~50% of the time Gareth and I spent pairing on this was trying to get the delete + create flow working right because it would break obscurely, so I'm a bit gun shy here ;-)

File: webapp/src/js/services/telemetry.js (hunk 12 above)
Comment (SCdF) [reply webapp/src/js/services/telemetry.js]: I am reasonably sure mango can't do reduce :-)

Comment (SCdF) [reply webapp/src/js/services/telemetry.js]: I don't think so. The core reason it's inline here is that because you only ever call it once before destroying the DB, I didn't see the need to create a design document just to query it and then delete it.

File: webapp/src/js/services/telemetry.js (hunk 17 above)
Comment (SCdF) [reply webapp/src/js/services/telemetry.js]: I was going to filter it to make it more explicit, but it's all vaguely useful. Things like the DB name, doc counts, update_seq etc. It depends on whether it's CouchDB or PouchDB.

File: webapp/src/js/services/telemetry.js (hunk 14 above)
Comment (SCdF) [reply webapp/src/js/services/telemetry.js]: Yeah I think that's a great call. It does add more work on the server side, but it's really just a 50-line script someone has to write and a design doc view on the meta DB (is that hard for some reason? Do we already have one?) to pull out `telemetry` docs.

Thoughts @garethbowen?

File: webapp/tests/karma/unit/services/telemetry.js (hunk 22 above)
Comment (SCdF) [reply webapp/tests/karma/unit/services/telemetry.js]: Yeah. The reasoning behind it being called "Record" is that we're already inside a describe called "Telemetry Service".

Is there a name you think would be clearer?

File: webapp/tests/karma/unit/services/telemetry.js (hunk 21 above)
Comment (SCdF) [reply webapp/tests/karma/unit/services/telemetry.js]: It's a fine line between strife and crime, as they say.

(I mean, IDK, I kind of dislike unit tests in general because they nearly always end up testing implementation details. You could argue this shows that we're retrieving the right things, at least sort of, though to truly test we'd have to check what each parameter was)

File: webapp/tests/karma/unit/services/telemetry.js (hunk 23 above)
Comment (SCdF) [reply webapp/tests/karma/unit/services/telemetry.js]: Yeah we could do that. There is `.to.deep.include` which lets you ignore things that you can't be bothered checking (like dates that are different all the time.)

File: webapp/tests/karma/unit/services/telemetry.js (hunk 24 above)
Comment (SCdF) [reply webapp/tests/karma/unit/services/telemetry.js]: No it's a good point, I'll remove the 409 reference and just talk about conflicts

File: webapp/tests/karma/unit/services/telemetry.js (hunk 22 above)
Comment (alxndrsn) [reply webapp/tests/karma/unit/services/telemetry.js]: for me, `#record()`, `.record()`, `record()`

File: webapp/src/js/controllers/inbox.js (hunk 27 above)
[the remark this answers is not in the document]
Comment (alxndrsn) [reply webapp/src/js/controllers/inbox.js]: I'm not fussed if it's implicit or explicit, either in the data model or Telemetry.record()` calls, but I think using a different divider for the namespace would:
* make it more obvious what the namespace was, e.g. `namespace.metric`, or `namespace:metric` if we're uncomfortable having `.` in a JSON property name
* make it possible to filter metrics by namespace when using simple tools like `grep`, `cut` etc.

File: webapp/src/js/services/telemetry.js (hunk 8 above)
Comment (alxndrsn) [reply webapp/src/js/services/telemetry.js]: > However, if you just log out of your session and log back in as someone else using the same browser, e.g. at a facility or something, that would be caught.

While this is an interesting scenario, perhaps not particularly interesting at least for load timings.  On the other hand, we currently have single-user-multi-login being used in prod projects with large datasets, where comparing performance and usage metrics for the same user, on the same device, but across his different accounts could be very interesting.  These users are very likely to have sllow startup times.

> We could use something from android, but then it wouldn't work outside of that, and ideally it would be consistent in shape (ie both should be UUIDs).

For metrics used for measuring performance, having a quick way to differentiate between android users and desktop users may be _very_ useful.  Also being able to differentiate between different apps on the same phone would be nice, for the scenario above.  Filtering by a combination of browser UserAgent and device ID would allow this.

My understanding is that getting a specific user ID for a desktop user isn't directly possible, but I don't think this means we should avoid doing it for mobile users.

> The core "real" problem we're solving here, is that if you are logged in on two different devices / browsers, then I don't want your aggregate telemetry documents to conflict with each other.
>
> Do you think sessionId sounds more correct?

If that's the aim, then using a random UUID sounds fine.  `sessionId` seems reasonable, as long as it's not already used elsewhere to mean something different.  I can only find one reference in the webapp repo, where `medic-api` refers to the `AuthSession` as `sessionId`, but this is an isolated case, and likely misleading in itself.

File: webapp/src/js/services/telemetry.js (hunk 9 above)
Comment (alxndrsn) [reply webapp/src/js/services/telemetry.js]: Sounds thorny.  Consider removing the UUID if you have the patience.  The only danger in leaving it here is we may fail to clean up the databases if we lose track of a UUID.

Comment (alxndrsn) [reply webapp/src/js/services/telemetry.js]: If it's kept, adding a comment to explain its presence would also be helpful.

File: webapp/src/js/services/telemetry.js (hunk 10 above)
Comment (alxndrsn) [reply webapp/src/js/services/telemetry.js]: Maybe you could `JSON.stringify()`, but perhaps then it's not really cleaning up the code :¬)

File: webapp/src/js/services/telemetry.js (hunk 15 above)
Comment (alxndrsn) [reply webapp/src/js/services/telemetry.js]: > To be super clear, we keep on reassigning `queue` on each call, so we are "dropping the head" from a memory allocation point of view.

:+1: I guess we just need to know if `Promise`s keep a reference to their parent.  A naive implementation probably would, but perhaps the implementation is smart enough not to.

File: webapp/src/js/services/telemetry.js (hunk 11 above)
Comment (garethbowen) [reply webapp/src/js/services/telemetry.js]: Initialising them in `aggregate()` as Alex suggested makes sense to me. I think I liked that the `LAST_AGGREGATED_DATE_KEY` was only ever used in this one function, but the name of the function doesn't make it clear that it also sets the value so it's a bit unexpected.

File: webapp/src/js/services/telemetry.js (hunk 14 above)
Comment (garethbowen) [reply webapp/src/js/services/telemetry.js]: We do already have a ddoc on the meta DB, but you might also be able to just query using all_docs because the doc _id is useful.

But yes I think it makes sense to store it in the user's meta db.

We need to think about how we're going to get this data out at some point...

File: webapp/tests/karma/unit/services/telemetry.js (hunk 20 above)
Comment (garethbowen) [reply webapp/tests/karma/unit/services/telemetry.js]: I'd like to keep unit tests as fast as possible so I think we should stick with mocking here. There would be some value in adding integration tests which could use a real pouchdb to ensure the units work together as expected.

File: webapp/src/js/services/telemetry.js (hunk 15 above)
Comment (garethbowen) [reply webapp/src/js/services/telemetry.js]: We ended up discussing this in the development channel but I don't know if we got any closer to an answer. I feel like we went with it as a quick fix for race conditions we were having. I think a different implementation would be easier to follow...

File: webapp/src/js/services/telemetry.js (hunk 9 above)
Comment (garethbowen) [reply webapp/src/js/services/telemetry.js]: I'm fairly sure we needed it because if you destroy a PouchDB instance then create a new one with the same name it fails saying the IDB is unavailable (or something).

**Comment by SCdF** [comment]
@alxndrsn / @garethbowen thanks for your notes chaps. I'm going to work on PouchDB stuff for now since bitrot on this branch seems unlikely (until we move to Angular I guess, and the service formate changes), and I'll pick it up again once I've delivered the first cut of that

**Comment by SCdF** [comment]
@garethbowen please double check again, I made some small changes but it's good to go

**Comment by garethbowen** [comment]
Awesome dude!
